src/server/twemcache-rs/src/admin.rs

```diff
@@ -0,0 +1,385 @@
+                        return;
+                    }
+                }
+            }
+
+            self.get_rusage();
+        }
+    }
+
+    /// Returns a `SyncSender` which can be used to send `Message`s to the
+    /// `Admin` component.
+    pub fn message_sender(&self) -> Sender<Message> {
+        self.message_sender.clone()
+    }
+
+    pub fn get_rusage(&self) {
+        let mut rusage = libc::rusage {
+            ru_utime: libc::timeval {
+                tv_sec: 0,
+                tv_usec: 0,
+            },
+            ru_stime: libc::timeval {
+                tv_sec: 0,
+                tv_usec: 0,
+            },
+            ru_maxrss: 0,
+            ru_ixrss: 0,
+            ru_idrss: 0,
+            ru_isrss: 0,
+            ru_minflt: 0,
+            ru_majflt: 0,
+            ru_nswap: 0,
+            ru_inblock: 0,
+            ru_oublock: 0,
+            ru_msgsnd: 0,
+            ru_msgrcv: 0,
+            ru_nsignals: 0,
+            ru_nvcsw: 0,
+            ru_nivcsw: 0,
+        };
+
+        if unsafe { libc::getrusage(libc::RUSAGE_SELF, &mut rusage) } == 0 {
+            set_counter!(
+                &Stat::RuUtime,
+                rusage.ru_utime.tv_sec as u64 * 1000000000 + rusage.ru_utime.tv_usec as u64 * 1000,
+            );
+            set_counter!(
+                &Stat::RuStime,
+                rusage.ru_stime.tv_sec as u64 * 1000000000 + rusage.ru_stime.tv_usec as u64 * 1000,
+            );
+            set_gauge!(&Stat::RuMaxrss, rusage.ru_maxrss as i64);
+            set_gauge!(&Stat::RuIxrss, rusage.ru_ixrss as i64);
+            set_gauge!(&Stat::RuIdrss, rusage.ru_idrss as i64);
+            set_gauge!(&Stat::RuIsrss, rusage.ru_isrss as i64);
+            set_counter!(&Stat::RuMinflt, rusage.ru_minflt as u64);
+            set_counter!(&Stat::RuMajflt, rusage.ru_majflt as u64);
+            set_counter!(&Stat::RuNswap, rusage.ru_nswap as u64);
+            set_counter!(&Stat::RuInblock, rusage.ru_inblock as u64);
+            set_counter!(&Stat::RuOublock, rusage.ru_oublock as u64);
+            set_counter!(&Stat::RuMsgsnd, rusage.ru_msgsnd as u64);
+            set_counter!(&Stat::RuMsgrcv, rusage.ru_msgrcv as u64);
+            set_counter!(&Stat::RuNsignals, rusage.ru_nsignals as u64);
+            set_counter!(&Stat::RuNvcsw, rusage.ru_nvcsw as u64);
+            set_counter!(&Stat::RuNivcsw, rusage.ru_nivcsw as u64);
+        }
+    }
+}
+
+impl EventLoop for Admin {
+    fn get_mut_session(&mut self, token: Token) -> Option<&mut Session> {
+        self.sessions.get_mut(token.0)
+    }
+
+    fn handle_data(&mut self, token: Token) -> Result<(), ()> {
+        trace!("handling request for admin session: {}", token.0);
+        if let Some(session) = self.sessions.get_mut(token.0) {
+            loop {
+                // TODO(bmartin): buffer should allow us to check remaining
+                // write capacity.
+                if session.write_pending() > (1024 - 6) {
```

Comment:
Please explain / name magic numbers

Reply:
It's very magic - I think this should be bumped up to an even KB, but the exact value is probably up for debate.

Reply:
Tied this to the initial buffer size. Should be better now.